core/client/fs/src/main/java/alluxio/client/block/stream/WorkerInternalClientDataReader.java

```diff
@@ -0,0 +1,158 @@
+  private final BlockReader mReader;
+  private final long mEnd;
+  private final long mChunkSize;
+  private long mPos;
+  private boolean mClosed;
+
+  /**
+   * Creates an instance of {@link WorkerInternalClientDataReader}.
+   *
+   * @param reader the block reader to read data from
+   * @param offset the offset
+   * @param len the length to read
+   * @param chunkSize the chunk size
+   */
+  private  WorkerInternalClientDataReader(BlockReader reader,
+      long offset, long len, long chunkSize) {
+    mReader = reader;
+    Preconditions.checkArgument(chunkSize > 0);
+    mPos = offset;
+    mEnd = Math.min(mReader.getLength(), offset + len);
+    mChunkSize = chunkSize;
+  }
+
+  @Override
+  public DataBuffer readChunk() throws IOException {
+    if (mPos >= mEnd) {
+      return null;
+    }
+    ByteBuffer buffer = mReader.read(mPos, Math.min(mChunkSize, mEnd - mPos));
+    DataBuffer dataBuffer = new NioDataBuffer(buffer, buffer.remaining());
+    mPos += dataBuffer.getLength();
+    MetricsSystem.counter(MetricKey.WORKER_BYTES_READ_DIRECT.getName()).inc(dataBuffer.getLength());
+    MetricsSystem.meter(MetricKey.WORKER_BYTES_READ_DIRECT_THROUGHPUT.getName())
+        .mark(dataBuffer.getLength());
+    return dataBuffer;
+  }
+
+  @Override
+  public long pos() {
+    return mPos;
+  }
+
+  @Override
+  public void close() {
+    if (mClosed) {
+      return;
+    }
+    mClosed = true;
+  }
+
+  /**
+   * Factory class to create {@link WorkerInternalClientDataReader}s.
+   */
+  @NotThreadSafe
+  public static class Factory implements DataReader.Factory {
+    private final WorkerInternalBlockWorker mWorkerInternalBlockWorker;
+    private final long mChunkSize;
+
+    private BlockReader mReader;
+    private boolean mClosed;
+    private ReadRequest mReadRequestPartial;
+    private BlockReadRequest mBlockReadRequest;
+
+    /**
+     * Creates an instance of {@link Factory}.
+     *
+     * @param context the file system context
+     * @param blockId the block ID
+     * @param chunkSize chunk size in bytes
+     * @param options the instream options
+     */
+    public Factory(FileSystemContext context, long blockId,
+        long chunkSize, InStreamOptions options)  {
+      mChunkSize = chunkSize;
+      mClosed = false;
+      boolean isPromote = ReadType
+          .fromProto(options.getOptions().getReadType()).isPromote();
+      mReadRequestPartial = ReadRequest.newBuilder()
+          .setBlockId(blockId).setPromote(isPromote).build();
+      mWorkerInternalBlockWorker = context.acquireWorkerInternalBlockWorkerClient();
+    }
+
+    @Override
+    public DataReader create(long offset, long len) throws IOException {
+      mReadRequestPartial = mReadRequestPartial.toBuilder()
+          .setOffset(offset).setLength(len).build();
+      mBlockReadRequest = new BlockReadRequest(mReadRequestPartial);
+      try {
+        mReader = mWorkerInternalBlockWorker.getBlockReader(mBlockReadRequest);
```

Comment:
is it necessary to create `mReadRequestPartial` in this Factory? I actually don't see how a proto is needed except to create  `mBlockReadRequest` to pass to `getBlockReader` and `cleanBlockReader`. If I recall correctly, we don't require a proto to be passed there, right?

Reply:
The BlockReadRequest is heavily bundled with Proto files:
```
public final class BlockReadRequest extends ReadRequest {
  private final Protocol.OpenUfsBlockOptions mOpenUfsBlockOptions;

  /**
   * Creates an instance of {@link BlockReadRequest}.
   *
   * @param request the block read request
   */
  public BlockReadRequest(alluxio.grpc.ReadRequest request) {
    super(request.getBlockId(), request.getOffset(), request.getOffset() + request.getLength(),
        request.getChunkSize());

    if (request.hasOpenUfsBlockOptions()) {
      mOpenUfsBlockOptions = request.getOpenUfsBlockOptions();
    } else {
      mOpenUfsBlockOptions = null;
    }
    mPromote = request.getPromote();
    mPositionShort = request.getPositionShort();
    // Note that we do not need to seek to offset since the block worker is created at the offset.
  }
```

I didn't fully move the proto logics out because I think that will impact more files and could be done later after merging all the fuse code back to the master branch if needed.